### ISSUE TITLE
Gracefully shutting down WebSocket connections on shutdown

### ISSUE THREAD

**Issue by samsp-msft** [microsoft/reverse-proxy#2622]
A request came in from a partner team, that they want YARP to gracefully drain and shutdown connections when the process gets a SIGTERM from k8s.
I think we drain existing inbound connections and not allow new ones to connect, but I don't know what we do upstream/downstream, especially with websocket and http2 connections?

**Comment by samsp-msft** [comment]
@timmydo for FYI.

**Comment by samsp-msft** [comment]
> The other point touched on by Sam was web sockets. Is there a way we could signal to YARP on ApplicationStopping that it tries to shutdown the existing websocket connections as gracefully as possible?

@MihaZupan ??

**Comment by MihaZupan** [comment]
The functionality doesn't exist in YARP yet, but it's possible to add if there's demand.

YARP currently treats WebSocket streams as opaque data streams, we just copy bytes back and forth.
Depending on how graceful you want to be, you could either close the underlying transport or try to parse the WebSocket stream to inject a close frame.

The second involves a bit more overhead to intercept the stream to keep track of frames, but it's relatively cheap, and YARP already has most of the logic for it in the [WS telemetry](https://github.com/microsoft/reverse-proxy/tree/main/src/ReverseProxy/WebSocketsTelemetry) implementation.

**Comment by samsp-msft** [comment]
@timmydo is this a formal ask for websockets closure?

**Comment by timmydo** [comment]
Yes please. Currently we have some custom code trying to track the HttpContext objects using web sockets. On shutdown we try to abort them all but we would prefer a softer close.
